application/common/models/Reset.php

```diff
@@ -337,4 +375,208 @@ public static function getExpireTimestamp()
+                $this->disable();
+            }
+        }
+    }
+
+    public function setType($type, $methodId = null)
+    {
+        $previousType = $this->type;
+        /*
+         * If type is not method, update or throw error
+         */
+        if (in_array($type, [self::TYPE_SPOUSE, self::TYPE_SUPERVISOR, self::TYPE_PRIMARY])) {
+            $this->type = $type;
+            $this->method_id = null;
+        } elseif ($type == self::TYPE_METHOD) {
+            /*
+             * If type is method but methodId is missing, throw error
+             */
+            if ($methodId === null) {
+                throw new BadRequestHttpException('Method ID required for reset type method', 1462988984);
+            }
+
+            /*
+             * Make sure user owns requested method before update
+             */
+            $method = Method::findOne(['id' => $methodId, 'user_id' => $this->user_id]);
+            if ($method === null) {
+                throw new NotFoundHttpException('Method not found', 1462989221);
+            }
+            $this->type = self::TYPE_METHOD;
+            $this->method_id = $methodId;
+        } else {
+            throw new BadRequestHttpException('Unknown reset type requested', 1462989489);
+        }
+
+        /*
+         * Save changes
+         */
+        $this->saveOrError('Set type of reset', 'Unable to update reset type.');
+
+        EventLog::log(
+            'ResetChangeType',
+            [
+                'reset_id' => $this->id,
+                'previous_type' => $previousType,
+                'new_type' => $this->type,
+                'attempts' => $this->attempts,
+            ],
+            $this->user_id
+        );
+    }
+
+    /**
+     * Increments attempts counter and disables account when limit is reached
+     * @param string $action Used in logging, either 'send' or 'verify'
+     * @throws ServerErrorHttpException
+     * @throws TooManyRequestsHttpException
+     */
+    public function trackAttempt($action)
+    {
+        /*
+         * Increment attempts count first thing
+         */
+        $this->attempts++;
+        $this->saveOrError($action . ' reset', 'Unable to increment attempts count.');
+
+        /*
+         * Enable / disable reset as needed
+         */
+        $this->enableOrDisableIfNeeded();
+
+        /*
+         * Check if reset is disabled and throw exception if it is
+         */
+        if ($this->isDisabled()) {
+            \Yii::error([
+                'action' => $action . ' reset',
+                'reset_id' => $this->id,
+                'attempts' => $this->attempts,
+                'status' => 'error',
+                'error' => 'Reset is currently disabled until ' . $this->disable_until,
+            ]);
+            throw new TooManyRequestsHttpException();
+        }
+    }
+
+    /**
+     * Save model or throw exception on error
+     * @param string $action
+     * @param string $errorPrefix
```

Comment:
Do you know if the `message` property of a `ServerErrorHttpException` is returned in the HTTP Response? If so, we might want to add a comment here in the doc. block that the value of `$errorPrefix` will probably be displayed to the user (just as a reminder to avoid putting anything sensitive in the value passed in for `$errorPrefix`). Just a thought.